src/metabase/mbql/normalize.clj

```diff
@@ -0,0 +1,541 @@
+(ns metabase.mbql.normalize
+  "Logic for taking any sort of weird MBQL query and normalizing it into a standardized, canonical form. You can think
+  of this like taking any 'valid' MBQL query and rewriting it as-if it was written in perfect up-to-date MBQL in the
+  latest version. There are two main things done here, done as three separate steps:
+
+  #### NORMALIZING TOKENS
+
+  Converting all identifiers to lower-case, lisp-case keywords. e.g. `{\"SOURCE_TABLE\" 10}` becomes `{:source-table
+  10}`.
+
+  #### CANONICALIZING THE QUERY
+
+  Rewriting deprecated MBQL 95 syntax and other things that are still supported for backwards-compatibility in
+  canonical MBQL 98 syntax. For example `{:breakout [:count 10]}` becomes `{:breakout [[:count [:field-id 10]]]}`.
+
+  #### REMOVING EMPTY CLAUSES
+
+  Removing empty clauses like `{:aggregation nil}` or `{:breakout []}`.
+
+  Token normalization occurs first, followed by canonicalization, followed by removing empty clauses."
+  (:require [clojure.walk :as walk]
+            [medley.core :as m]
+            [metabase.mbql.util :as mbql.u]
+            [metabase.util :as u]
+            [puppetlabs.i18n.core :refer [tru]]))
```

Comment:
`metabase.util.i18n` might not have existed when you started writing this, but would be good to get it switched to that.

Reply:
yeah I noticed there are still several namespaces that use the puppet i18n namespace I wasn't sure what the rule was when to use which